### PR TITLE
[WGSL] texture_external as a global generates incorrect indices

### PR DESCRIPTION
**Pull request by tadeuzagallo**
#### e1551b11d288edb551694c530411a026b42c7747
<pre>
[WGSL] texture_external as a global generates incorrect indices
<a href="https://bugs.webkit.org/show_bug.cgi?id=263564">https://bugs.webkit.org/show_bug.cgi?id=263564</a>
rdar://117379167

Reviewed by Mike Wyrzykowski.

The generated auto layout shouldn&apos;t use the user-provided @binding values directly,
but instead generate new binding IDs. This solves the problem for texture_external
as well as generates contiguous IDs. The code can be simplified further, but for
now it exposes extra information for the API about the offsets of the bufferSize
special entries, while keeping the current behavior working. Once we update the API
we can simplify the code without breaking anything.

* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::determineUsedGlobals):
(WGSL::RewriteGlobalVariables::insertStructs):
* Source/WebGPU/WGSL/tests/valid/texture-external.wgsl: Added.

(WGSL::RewriteGlobalVariables::insertStructs):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::Device::createBindGroupLayout):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::addPipelineLayouts):
(WebGPU::Device::generatePipelineLayout):
Update BindGroupLayout generation to match the compiler change.

Canonical link: <a href="https://commits.webkit.org/269996@main">https://commits.webkit.org/269996@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d321dde710452da4ef3179d90c34f9f317f5d38f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24218 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26350 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22296 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3956 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24685 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24461 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1850 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20930 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26939 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1595 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28075 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22080 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/22149 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/19189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/2464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5805 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1935 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1898 "Built successfully") | | |

### CODE REVIEW

**Review by mwyrzykowski — DISMISSED** [review]

**Review by mwyrzykowski — COMMENTED** [review]
This appears to break https://webgpu.github.io/webgpu-samples/samples/texturedCube

File: Source/WebGPU/WGSL/GlobalVariableRewriter.cpp [hunk 1]
```diff
@@ -1092,7 +1082,38 @@ Vector<unsigned> RewriteGlobalVariables::insertStructs(const UsedResources& used
...
             ASSERT(global.declaration->maybeTypeName());
 
-            entries.append({ binding, &createArgumentBufferEntry(binding, *global.declaration) });
+            entries.append({ metalId, &createArgumentBufferEntry(metalId, *global.declaration) });
+
+            if (m_generatedLayout->bindGroupLayouts.size() <= group)
+                m_generatedLayout->bindGroupLayouts.grow(group + 1);
+
+            BindGroupLayoutEntry entry {
+                .binding = metalId,
```
Comment (mwyrzykowski) [inline Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:1091]: do we need to take into account `global.resource->binding` here or does this not matter based on the specification wording?

**Review by mwyrzykowski — COMMENTED** [review]
It looks like this causes a regression because the BindGroup is still using the old indices. Like from the textured-cube GitHub sample:
```
  const uniformBindGroup = device.createBindGroup({
    layout: pipeline.getBindGroupLayout(0),
    entries: [
      {
        binding: 0,
        resource: {
          buffer: uniformBuffer,
        },
      },
      {
        binding: 1,
        resource: sampler,
      },
      {
        binding: 2,
        resource: cubeTexture.createView(),
      },
    ],
  });
```

**Comment by mwyrzykowski** [comment]
I'm going to write up a patch to fix up the BindGroupLayout and BindGroup and then we can submit the patch with this PR to avoid any regressions

**Comment by mwyrzykowski** [comment]
Can you please apply this patch to the PR prior to merging? It should fix the regressions with the bind group layouts:
[patch.patch](https://github.com/WebKit/WebKit/files/13201469/patch.patch)

**Review by mwyrzykowski — APPROVED** [review]
Approval conditional on applying the attached patch so we don't regress the CTS tests

[patch.patch](https://github.com/WebKit/WebKit/files/13201487/patch.patch)

**Comment by Ahmad-S792** [comment]
@tadeuzagallo - Need to be rebased to resolve conflict.

**Pull request merged**